File: node/src/a2a.ts
import { inspect } from 'node:util';

import { JacsModuleError } from './integrations.js';

type AnyFunction = (...args: any[]) => any;
type AnyModule = Record<string, unknown>;

export type A2ATrustPolicy = 'open' | 'verified' | 'strict';

export interface GetA2AIntegrationOptions {
  trustPolicy?: A2ATrustPolicy;
}

export interface QuickstartA2AOptions {
  name: string;
  domain: string;
  description: string;
  algorithm?: string;
  configPath?: string;
  url?: string;
}

export interface RegisterWithAgentCardOptions extends GetA2AIntegrationOptions {
  ownerEmail?: string;
  domain?: string;
  description?: string;
  agentJson?: string | Record<string, unknown>;
  publicKeyPem?: string;
}

export interface RegisterWithAgentCardResult {
  registration: unknown;
  agentCard: Record<string, unknown>;
  agentJson: string;
}

export interface A2AMediatedJobOptions extends GetA2AIntegrationOptions {
  transport?: 'sse' | 'ws';
  verifyInboundArtifact?: boolean;
  enforceTrustPolicy?: boolean;
  maxReconnectAttempts?: number;
  notifyEmail?: string;
  emailSubject?: string;
}

function isMissingModuleError(error: unknown, moduleName: string): boolean {
  if (!error || typeof error !== 'object') {
    return false;
  }

  const code = (error as { code?: string }).code;
  if (code === 'ERR_MODULE_NOT_FOUND' || code === 'MODULE_NOT_FOUND') {
    return true;
  }

  const message = (error as { message?: string }).message ?? '';
  if (message.includes(`Cannot find package '${moduleName}'`)) {
    return true;
  }
  if (message.includes(`Cannot find module '${moduleName}'`)) {
    return true;
  }
  if (message.includes(`Failed to load url ${moduleName}`)) {
    return true;
  }
  return false;
}

async function loadOptionalModule(
  moduleName: string,
  feature: string,
  installHint: string,
): Promise<AnyModule> {
  try {
    return await import(moduleName) as AnyModule;
  } catch (error) {
    if (isMissingModuleError(error, moduleName)) {
      throw new JacsModuleError(
        `Optional dependency '${moduleName}' is required for ${feature}. ${installHint}`,
      );
    }
    throw error;
  }
}

function getRequiredFunction(moduleName: string, moduleObj: AnyModule, name: string): AnyFunction {
  const value = moduleObj[name];
  if (typeof value !== 'function') {
    throw new JacsModuleError(
      `Module '${moduleName}' does not export function '${name}'. ` +
      `Received: ${inspect(value, { depth: 1 })}`,
    );
  }
  return value as AnyFunction;
}

function getRequiredClass(moduleName: string, moduleObj: AnyModule, name: string): new (...args: any[]) => any {
  const value = moduleObj[name];
  if (typeof value !== 'function') {
    throw new JacsModuleError(
      `Module '${moduleName}' does not export class '${name}'. ` +
      `Received: ${inspect(value, { depth: 1 })}`,
    );
  }
  return value as new (...args: any[]) => any;
}

const A2A_HINT = "Install with: npm install @hai.ai/jacs";
const A2A_MODULE = '@hai.ai/jacs/a2a';

async function callIntegrationMethod(
  integration: Record<string, unknown>,
  methodName: string,
  args: unknown[],
): Promise<unknown> {
  const method = integration[methodName];
  if (typeof method !== 'function') {
    throw new JacsModuleError(
      `A2A integration missing method '${methodName}'. Received: ${inspect(method, { depth: 1 })}`,
    );
  }
  return (method as AnyFunction).apply(integration, args);
}

function asRecord(value: unknown, label: string): Record<string, unknown> {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    throw new JacsModuleError(`${label} must be an object`);
  }
  return value as Record<string, unknown>;
}

function stringifyAgentJson(agentJson: string | Record<string, unknown>): string {
  if (typeof agentJson === 'string') {
    return agentJson;
  }
  return JSON.stringify(agentJson);
}

function resolveCardProfile(agentCard: Record<string, unknown>): string {
  const metadata = agentCard.metadata;
  if (metadata && typeof metadata === 'object') {
    const profile = (metadata as Record<string, unknown>).a2aProfile;
    if (typeof profile === 'string' && profile.trim() !== '') {
      return profile;
    }
  }

  const protocolVersions = agentCard.protocolVersions;
  if (Array.isArray(protocolVersions) && typeof protocolVersions[0] === 'string' && protocolVersions[0].trim() !== '') {
    return protocolVersions[0];
  }

  const supportedInterfaces = agentCard.supportedInterfaces;
  if (Array.isArray(supportedInterfaces)) {
    for (const entry of supportedInterfaces) {
      if (!entry || typeof entry !== 'object') continue;
      const version = (entry as Record<string, unknown>).protocolVersion;
      if (typeof version === 'string' && version.trim() !== '') {
        return version;
      }
    }
  }

  return '0.4.0';
}

function valueAsString(record: Record<string, unknown>, key: string): string {
  const value = record[key];
  return typeof value === 'string' ? value : '';
}

function mergeOptions<T extends Record<string, unknown>>(value: T): T {
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(value)) {
    if (v !== undefined) {
      out[k] = v;
    }
  }
  return out as T;
}

function assertRequiredIdentityField(value: unknown, fieldName: string): asserts value is string {
  if (typeof value !== 'string' || value.trim().length === 0) {
    throw new JacsModuleError(`quickstartA2A requires non-empty '${fieldName}'`);
  }
}

/**
 * Build a JACS A2A integration instance from a caller-provided JACS client.
 *
 * The JACS client is typically an instance of `JacsClient` from `@hai.ai/jacs/client`.
 */
export async function getA2AIntegration(
  jacsClient: unknown,
  options: GetA2AIntegrationOptions = {},
): Promise<unknown> {
  const mod = await loadOptionalModule(A2A_MODULE, 'A2A integration', A2A_HINT);
  const IntegrationClass = getRequiredClass(A2A_MODULE, mod, 'JACSA2AIntegration');
  return new IntegrationClass(jacsClient, options.trustPolicy);
}

/**
 * Delegate to `JACSA2AIntegration.quickstart(...)` from `@hai.ai/jacs/a2a`.
 */
export async function quickstartA2A(options: QuickstartA2AOptions): Promise<unknown> {
  assertRequiredIdentityField(options.name, 'name');
  assertRequiredIdentityField(options.domain, 'domain');
  assertRequiredIdentityField(options.description, 'description');

  const mod = await loadOptionalModule(A2A_MODULE, 'A2A quickstart', A2A_HINT);
  const IntegrationClass = getRequiredClass(A2A_MODULE, mod, 'JACSA2AIntegration');
  const quickstart = getRequiredFunction(A2A_MODULE, IntegrationClass as unknown as AnyModule, 'quickstart');
  return quickstart(mergeOptions({
    ...options,
    algorithm: options.algorithm ?? 'pq2025',
  }));
}

export async function exportAgentCard(
  jacsClient: unknown,
  agentData: Record<string, unknown>,
  options: GetA2AIntegrationOptions = {},
): Promise<unknown> {
  const integration = await getA2AIntegration(jacsClient, options) as Record<string, unknown>;
  return callIntegrationMethod(integration, 'exportAgentCard', [agentData]);
}

export async function signArtifact(
  jacsClient: unknown,
  artifact: Record<string, unknown>,
  artifactType: string,
  parentSignatures: Record<string, unknown>[] | null = null,
  options: GetA2AIntegrationOptions = {},
): Promise<unknown> {
  const integration = await getA2AIntegration(jacsClient, options) as Record<string, unknown>;
  return callIntegrationMethod(integration, 'signArtifact', [artifact, artifactType, parentSignatures]);
}

export async function verifyArtifact(
  jacsClient: unknown,
  wrappedArtifact: string | Record<string, unknown>,
  options: GetA2AIntegrationOptions = {},
): Promise<unknown> {
  const integration = await getA2AIntegration(jacsClient, options) as Record<string, unknown>;
  return callIntegrationMethod(integration, 'verifyWrappedArtifact', [wrappedArtifact]);
}

export async function createChainOfCustody(
  jacsClient: unknown,
  artifacts: Record<string, unknown>[],
  options: GetA2AIntegrationOptions = {},
): Promise<unknown> {
  const integration = await getA2AIntegration(jacsClient, options) as Record<string, unknown>;
  return callIntegrationMethod(integration, 'createChainOfCustody', [artifacts]);
}

export async function generateWellKnownDocuments(
  jacsClient: unknown,
  agentCard: unknown,
  jwsSignature: string,
  publicKeyB64: string,
  agentData: Record<string, unknown>,
  options: GetA2AIntegrationOptions = {},
): Promise<unknown> {
  const integration = await getA2AIntegration(jacsClient, options) as Record<string, unknown>;
  return callIntegrationMethod(
    integration,
    'generateWellKnownDocuments',
    [agentCard, jwsSignature, publicKeyB64, agentData],
  );
}

export async function assessRemoteAgent(
  jacsClient: unknown,
  agentCardJson: string | Record<string, unknown>,
  options: GetA2AIntegrationOptions = {},
): Promise<unknown> {
  const integration = await getA2AIntegration(jacsClient, options) as Record<string, unknown>;
  return callIntegrationMethod(integration, 'assessRemoteAgent', [agentCardJson]);
}

export async function trustA2AAgent(
  jacsClient: unknown,
  agentCardJson: string | Record<string, unknown>,
  options: GetA2AIntegrationOptions = {},
): Promise<unknown> {
  const integration = await getA2AIntegration(jacsClient, options) as Record<string, unknown>;
  return callIntegrationMethod(integration, 'trustA2AAgent', [agentCardJson]);
}

export function mergeAgentJsonWithAgentCard(
  agentJson: string | Record<string, unknown>,
  agentCard: Record<string, unknown>,
): string {
  const base = typeof agentJson === 'string'
    ? JSON.parse(agentJson) as Record<string, unknown>
    : { ...agentJson };
  const card = asRecord(agentCard, 'agentCard');

  base.a2aAgentCard = card;
  if (base.skills === undefined && Array.isArray(card.skills)) {
    base.skills = card.skills;
  }
  if (base.capabilities === undefined && card.capabilities && typeof card.capabilities === 'object') {
    base.capabilities = card.capabilities;
  }

  const metadata = (base.metadata && typeof base.metadata === 'object')
    ? { ...(base.metadata as Record<string, unknown>) }
    : {};
  metadata.a2aProfile = resolveCardProfile(card);
  metadata.a2aSkillsCount = Array.isArray(card.skills) ? card.skills.length : 0;
  base.metadata = metadata;

  return JSON.stringify(base);
}

export async function registerWithAgentCard(
  haiClient: unknown,
  jacsClient: unknown,
  agentData: Record<string, unknown>,
  options: RegisterWithAgentCardOptions = {},
): Promise<RegisterWithAgentCardResult> {
  const clientRecord = asRecord(haiClient, 'haiClient');
  const register = getRequiredFunction('HaiClient', clientRecord, 'register');
  const exportKeys = clientRecord.exportKeys;

  const cardRaw = await exportAgentCard(jacsClient, agentData, options);
  const agentCard = asRecord(cardRaw, 'agentCard');

  let baseAgentJson: string | Record<string, unknown>;
  if (options.agentJson !== undefined) {
    baseAgentJson = options.agentJson;
  } else {
    baseAgentJson = {
      jacsId: valueAsString(agentData, 'jacsId') || valueAsString(clientRecord, 'jacsId'),
      name: valueAsString(agentData, 'jacsName') || valueAsString(clientRecord, 'agentName'),
      jacsVersion: valueAsString(agentData, 'jacsVersion') || '1.0.0',
    };
  }
  const mergedAgentJson = mergeAgentJsonWithAgentCard(baseAgentJson, agentCard);

  let publicKeyPem = options.publicKeyPem;
  if (!publicKeyPem && typeof exportKeys === 'function') {
    const exported = exportKeys.apply(clientRecord);
    if (exported && typeof exported === 'object') {
      const candidate = (exported as Record<string, unknown>).publicKeyPem;
      if (typeof candidate === 'string' && candidate.length > 0) {
        publicKeyPem = candidate;
      }
    }
  }

  const registration = await register.apply(clientRecord, [mergeOptions({
    ownerEmail: options.ownerEmail,
    domain: options.domain,
    description: options.description,
    agentJson: mergedAgentJson,
    publicKeyPem,
  })]);

  return {
    registration,
    agentCard,
    agentJson: mergedAgentJson,
  };
}

export async function onMediatedBenchmarkJob(
  haiClient: unknown,
  jacsClient: unknown,
  handler: (taskArtifact: Record<string, unknown>) => Promise<Record<string, unknown>>,
  options: A2AMediatedJobOptions = {},
): Promise<void> {
  const clientRecord = asRecord(haiClient, 'haiClient');
  const onBenchmarkJob = getRequiredFunction('HaiClient', clientRecord, 'onBenchmarkJob');
  const submitResponse = getRequiredFunction('HaiClient', clientRecord, 'submitResponse');
  const sendEmail = clientRecord.sendEmail;

  const verifyInboundArtifact = options.verifyInboundArtifact ?? false;
  const enforceTrustPolicy = options.enforceTrustPolicy ?? false;
  const maxReconnectAttempts = options.maxReconnectAttempts ?? 0;
  const transport = options.transport ?? 'sse';

  let attempts = 0;
  while (true) {
    try {
      await onBenchmarkJob.apply(clientRecord, [async (job: Record<string, unknown>) => {
        const data = asRecord(job.data ?? {}, 'benchmark job data');
        const jobId = valueAsString(data, 'job_id') || valueAsString(data, 'run_id') || valueAsString(job, 'runId');
        if (!jobId) {
          throw new JacsModuleError('benchmark job missing job_id/run_id');
        }

        if (enforceTrustPolicy) {
          const remoteCard = data.remoteAgentCard ?? data.remote_agent_card ?? data.agentCard;
          if (!remoteCard) {
            throw new JacsModuleError('remote agent card required when trust enforcement is enabled');
          }
          const trust = await assessRemoteAgent(jacsClient, remoteCard as string | Record<string, unknown>, options) as Record<string, unknown>;
          if (trust.allowed !== true) {
            const reason = typeof trust.reason === 'string' ? trust.reason : 'unknown reason';
            throw new JacsModuleError(`trust policy rejected remote agent: ${reason}`);
          }
        }

        if (verifyInboundArtifact) {
          const inbound = data.a2aTask ?? data.a2a_task;
          if (!inbound) {
            throw new JacsModuleError('inbound a2a task required when signature verification is enabled');
          }
          const verification = await verifyArtifact(jacsClient, inbound as string | Record<string, unknown>, options) as Record<string, unknown>;
          if (verification.valid !== true) {
            const err = typeof verification.error === 'string' ? verification.error : 'unknown verification failure';
            throw new JacsModuleError(`inbound a2a task signature invalid: ${err}`);
          }
        }

        const taskPayload: Record<string, unknown> = {
          type: 'benchmark_job',
          jobId,
          scenarioId: data.scenario_id ?? data.scenarioId ?? null,
          config: data.config ?? null,
        };
        const taskArtifact = await signArtifact(jacsClient, taskPayload, 'task', null, options) as Record<string, unknown>;
        const resultPayload = await handler(taskArtifact);
        const resultArtifact = await signArtifact(
          jacsClient,
          resultPayload,
          'task-result',
          [taskArtifact],
          options,
        ) as Record<string, unknown>;

        const messageValue = resultPayload.message;
        const message = typeof messageValue === 'string' ? messageValue : JSON.stringify(resultPayload);

        await submitResponse.apply(clientRecord, [jobId, message, {
          metadata: {
            a2aTask: taskArtifact,
            a2aResult: resultArtifact,
          },
          processingTimeMs: 0,
        }]);

        if (options.notifyEmail && typeof sendEmail === 'function') {
          const subject = options.emailSubject ?? `A2A mediated result for job ${jobId}`;
          await (sendEmail as AnyFunction).apply(clientRecord, [{
            to: options.notifyEmail,
            subject,
            body: `Signed A2A artifact:\n\n${JSON.stringify(resultArtifact, null, 2)}`,
          }]);
        }
      }, { transport }]);
      return;
    } catch (error) {
      if (attempts >= maxReconnectAttempts) {
        throw error;
      }
      attempts += 1;
    }
  }
}

File: python/src/haisdk/a2a.py
"""A2A integration wrappers for HAISDK.

These helpers expose JACS A2A functionality from the `haisdk` namespace while
keeping implementation ownership in `jacs`.
"""

from __future__ import annotations

import json
from typing import Any

from haisdk._optional import load_optional_module, require_attr


def _get_integration_class() -> type:
    module = load_optional_module(
        "jacs.a2a",
        feature="A2A integration",
        install_hint='Install with: pip install "haisdk[a2a]"',
    )
    return require_attr(
        module,
        "JACSA2AIntegration",
        owner_name="jacs.a2a",
        upgrade_hint="Install/upgrade JACS with A2A support: pip install -U 'jacs[a2a]'",
    )


def get_a2a_integration(client: Any, trust_policy: str = "verified") -> Any:
    """Create a JACS A2A integration bound to a caller-provided JACS client."""
    cls = _get_integration_class()
    return cls(client, trust_policy=trust_policy)


def quickstart_a2a(
    name: str,
    domain: str,
    description: str,
    algorithm: str = "pq2025",
    config_path: str | None = None,
    url: str | None = None,
) -> Any:
    """Delegate to ``JACSA2AIntegration.quickstart(...)``."""
    cls = _get_integration_class()
    quickstart = require_attr(
        cls,
        "quickstart",
        owner_name="jacs.a2a.JACSA2AIntegration",
        upgrade_hint="Install/upgrade JACS with A2A support: pip install -U 'jacs[a2a]'",
    )
    if not callable(quickstart):
        raise ImportError(
            "jacs.a2a.JACSA2AIntegration.quickstart is missing. "
            "Install/upgrade JACS with A2A support: pip install -U 'jacs[a2a]'"
        )
    return quickstart(
        name=name,
        domain=domain,
        description=description,
        algorithm=algorithm,
        config_path=config_path,
        url=url,
    )


def export_agent_card(
    client: Any,
    agent_data: dict[str, Any],
    trust_policy: str = "verified",
) -> Any:
    integration = get_a2a_integration(client, trust_policy=trust_policy)
    return integration.export_agent_card(agent_data)


def sign_artifact(
    client: Any,
    artifact: dict[str, Any],
    artifact_type: str,
    parent_signatures: list[dict[str, Any]] | None = None,
    trust_policy: str = "verified",
) -> Any:
    integration = get_a2a_integration(client, trust_policy=trust_policy)
    return integration.sign_artifact(artifact, artifact_type, parent_signatures)


def verify_artifact(
    client: Any,
    wrapped_artifact: dict[str, Any],
    assess_trust: bool = False,
    policy: str | None = None,
    trust_policy: str = "verified",
) -> Any:
    integration = get_a2a_integration(client, trust_policy=trust_policy)
    return integration.verify_wrapped_artifact(
        wrapped_artifact,
        assess_trust=assess_trust,
        trust_policy=policy,
    )


def create_chain_of_custody(
    client: Any,
    artifacts: list[dict[str, Any]],
    trust_policy: str = "verified",
) -> Any:
    integration = get_a2a_integration(client, trust_policy=trust_policy)
    return integration.create_chain_of_custody(artifacts)


def generate_well_known_documents(
    client: Any,
    agent_card: Any,
    jws_signature: str,
    public_key_b64: str,
    agent_data: dict[str, Any],
    trust_policy: str = "verified",
) -> Any:
    integration = get_a2a_integration(client, trust_policy=trust_policy)
    return integration.generate_well_known_documents(
        agent_card,
        jws_signature,
        public_key_b64,
        agent_data,
    )


def assess_remote_agent(
    client: Any,
    agent_card_json: str,
    policy: str | None = None,
    trust_policy: str = "verified",
) -> Any:
    integration = get_a2a_integration(client, trust_policy=trust_policy)
    return integration.assess_remote_agent(agent_card_json, policy=policy)


def trust_a2a_agent(
    client: Any,
    agent_card_json: str,
    trust_policy: str = "verified",
) -> Any:
    integration = get_a2a_integration(client, trust_policy=trust_policy)
    return integration.trust_a2a_agent(agent_card_json)


def _resolve_card_profile(agent_card: dict[str, Any]) -> str:
    metadata = agent_card.get("metadata")
    if isinstance(metadata, dict):
        profile = metadata.get("a2aProfile")
        if isinstance(profile, str) and profile.strip():
            return profile

    versions = agent_card.get("protocolVersions")
    if isinstance(versions, list) and versions:
        first = versions[0]
        if isinstance(first, str) and first.strip():
            return first

    interfaces = agent_card.get("supportedInterfaces")
    if isinstance(interfaces, list):
        for interface in interfaces:
            if not isinstance(interface, dict):
                continue
            version = interface.get("protocolVersion")
            if isinstance(version, str) and version.strip():
                return version

    return "0.4.0"


def merge_agent_json_with_agent_card(
    agent_json: str | dict[str, Any],
    agent_card: dict[str, Any],
) -> str:
    if isinstance(agent_json, str):
        base = json.loads(agent_json)
    else:
        base = dict(agent_json)

    if not isinstance(base, dict):
        raise ValueError("agent_json must decode to an object")
    if not isinstance(agent_card, dict):
        raise ValueError("agent_card must be an object")

    base["a2aAgentCard"] = agent_card
    if "skills" not in base and isinstance(agent_card.get("skills"), list):
        base["skills"] = agent_card["skills"]
    if "capabilities" not in base and isinstance(agent_card.get("capabilities"), dict):
        base["capabilities"] = agent_card["capabilities"]

    metadata = base.get("metadata")
    if not isinstance(metadata, dict):
        metadata = {}
    metadata["a2aProfile"] = _resolve_card_profile(agent_card)
    metadata["a2aSkillsCount"] = (
        len(agent_card.get("skills", []))
        if isinstance(agent_card.get("skills"), list)
        else 0
    )
    base["metadata"] = metadata
    return json.dumps(base)


def register_with_agent_card(
    hai_client: Any,
    jacs_client: Any,
    hai_url: str,
    agent_data: dict[str, Any],
    *,
    owner_email: str | None = None,
    domain: str | None = None,
    description: str | None = None,
    agent_json: str | dict[str, Any] | None = None,
    public_key: str | None = None,
    trust_policy: str = "verified",
) -> dict[str, Any]:
    card = export_agent_card(jacs_client, agent_data, trust_policy=trust_policy)

    base_agent_json: str | dict[str, Any]
    if agent_json is not None:
        base_agent_json = agent_json
    else:
        base_agent_json = {
            "jacsId": agent_data.get("jacsId"),
            "name": agent_data.get("jacsName"),
            "jacsVersion": agent_data.get("jacsVersion", "1.0.0"),
        }

    merged_agent_json = merge_agent_json_with_agent_card(base_agent_json, card)
    registration = hai_client.register(
        hai_url,
        agent_json=merged_agent_json,
        public_key=public_key,
        owner_email=owner_email,
    )
    return {
        "registration": registration,
        "agent_card": card,
        "agent_json": merged_agent_json,
        "domain": domain,
        "description": description,
    }


def on_mediated_benchmark_job(
    hai_client: Any,
    jacs_client: Any,
    hai_url: str,
    handler: Any,
    *,
    transport: str = "sse",
    trust_policy: str = "verified",
    verify_inbound_artifact: bool = False,
    enforce_trust_policy: bool = False,
    max_reconnect_attempts: int = 0,
    notify_email: str | None = None,
    email_subject: str | None = None,
) -> None:
    attempts = 0
    while True:
        try:
            for event in hai_client.connect(hai_url, transport=transport):
                event_type = getattr(event, "event_type", None)
                if event_type != "benchmark_job":
                    continue

                data = getattr(event, "data", {})
                if not isinstance(data, dict):
                    data = {}

                job_id = (
                    data.get("job_id")
                    or data.get("run_id")
                    or data.get("jobId")
                    or data.get("runId")
                )
                if not isinstance(job_id, str) or not job_id:
                    raise ValueError("benchmark job missing job_id/run_id")

                if enforce_trust_policy:
                    remote_card = (
                        data.get("remoteAgentCard")
                        or data.get("remote_agent_card")
                        or data.get("agentCard")
                    )
                    if remote_card is None:
                        raise ValueError(
                            "remote agent card required when trust enforcement is enabled"
                        )
                    remote_card_json = (
                        remote_card
                        if isinstance(remote_card, str)
                        else json.dumps(remote_card)
                    )
                    trust = assess_remote_agent(
                        jacs_client,
                        remote_card_json,
                        trust_policy=trust_policy,
                    )
                    if not trust.get("allowed", False):
                        raise ValueError(
                            f"trust policy rejected remote agent: {trust.get('reason', 'unknown')}"
                        )

                if verify_inbound_artifact:
                    inbound = data.get("a2aTask") or data.get("a2a_task")
                    if inbound is None:
                        raise ValueError(
                            "inbound a2a task required when signature verification is enabled"
                        )
                    verify = verify_artifact(
                        jacs_client,
                        inbound if isinstance(inbound, dict) else json.loads(str(inbound)),
                        trust_policy=trust_policy,
                    )
                    if not verify.get("valid", False):
                        raise ValueError(
                            "inbound a2a task signature invalid: "
                            + str(verify.get("error", "unknown verification failure"))
                        )

                task_payload = {
                    "type": "benchmark_job",
                    "jobId": job_id,
                    "scenarioId": data.get("scenario_id") or data.get("scenarioId"),
                    "config": data.get("config"),
                }
                task_artifact = sign_artifact(
                    jacs_client,
                    task_payload,
                    "task",
                    trust_policy=trust_policy,
                )
                result_payload = handler(task_artifact)
                if not isinstance(result_payload, dict):
                    raise ValueError("handler must return a dict payload")

                result_artifact = sign_artifact(
                    jacs_client,
                    result_payload,
                    "task-result",
                    parent_signatures=[task_artifact],
                    trust_policy=trust_policy,
                )

                message = result_payload.get("message")
                if not isinstance(message, str):
                    message = json.dumps(result_payload)

                hai_client.submit_response(
                    hai_url,
                    job_id=job_id,
                    message=message,
                    metadata={
                        "a2aTask": task_artifact,
                        "a2aResult": result_artifact,
                    },
                    processing_time_ms=0,
                )

                if notify_email:
                    hai_client.send_email(
                        hai_url,
                        to=notify_email,
                        subject=email_subject or f"A2A mediated result for job {job_id}",
                        body="Signed A2A artifact:\n\n"
                        + json.dumps(result_artifact, indent=2),
                    )
            return
        except Exception:
            if attempts >= max_reconnect_attempts:
                raise
            attempts += 1


def __getattr__(name: str) -> Any:
    """Lazy passthrough so stable `jacs.a2a` symbols are reachable via `haisdk.a2a`."""
    module = load_optional_module(
        "jacs.a2a",
        feature="A2A integration",
        install_hint='Install with: pip install "haisdk[a2a]"',
    )
    try:
        return getattr(module, name)
    except AttributeError as exc:
        raise AttributeError(f"module 'haisdk.a2a' has no attribute {name!r}") from exc


__all__ = [
    "get_a2a_integration",
    "quickstart_a2a",
    "export_agent_card",
    "sign_artifact",
    "verify_artifact",
    "create_chain_of_custody",
    "generate_well_known_documents",
    "assess_remote_agent",
    "trust_a2a_agent",
    "merge_agent_json_with_agent_card",
    "register_with_agent_card",
    "on_mediated_benchmark_job",
]
